lib/flow/VaporField.cpp

```diff
@@ -1,37 +1,179 @@
-#include <sstream>
-
 #include "vapor/VaporField.h"
 #include "vapor/ConstantGrid.h"
 #include "GrownGrid.h"
 
+
 using namespace flow;
 
-// Constructor
+
+//
+// Class GridKey
+//
+void GridKey::Reset( uint64_t ts, int32_t ref, int32_t comp, std::string var,
+                     const std::vector<double>& min,
+                     const std::vector<double>& max,
+                     float dz )
+{
+    timestep = ts;
+    refLev   = ref;
+    compLev  = comp;
+    defaultZ = dz;
+
+    for( int i = 0; i < 3; i++ ) {
+        ext_min[i] = 0.0;
+        ext_max[i] = 0.0;
+    }
+    for( int i = 0; i < min.size(); i++ )
+        ext_min[i] = min[i];
+    for( int i = 0; i < max.size(); i++ )
+        ext_max[i] = max[i];
+
+    varName  = std::move( var );
+}
+
+bool GridKey::operator==(const GridKey& other) const
+{
+    // String comparison seems to occur most frequently, so put it at the first.
+    if( this->varName  != other.varName )
+        return false;
+    if( this->timestep != other.timestep )
+        return false;
+    if( this->refLev   != other.refLev )
+        return false;
+    if( this->compLev  != other.compLev )
+        return false;
+    if( this->defaultZ != other.defaultZ)
+        return false;
+    for( int i = 0; i < 3; i++ ) {
```

Comment:
Just an FYI that std::vector (and every other container I'm aware of) supports a '==' operator. I.e. you can just say:

`if( this->ext_min != other.ext_min )`

and avoid the element-wise comparison

Reply:
Well, I store the 3-element structure as a std::array instead of a std::vector, thus cannot use the built in == operator.

The reason to not use std::vector is to avoid 2 heap memory allocations, each 3-element in size. This decision was part of the effort to optimize grid retrieval from the grid cache.

Reply:
Are you sure std::array doesn't support '=='? It looks like it does:

https://en.cppreference.com/w/cpp/container/array

Reply:
You're right. I thought I was comparing a vector and an array. Now it's fixed, thanks for the reminder ;)